protocol-units/da-sequencer/node/src/storage/mod.rs

```diff
@@ -43,7 +131,79 @@ impl Storage {
 	pub fn produce_next_block(
 		&self,
 	) -> std::result::Result<Option<SequencerBlock>, DaSequencerError> {
-		todo!();
+		// Step 1: Determine the next block height
+		let next_height = self.determine_next_block_height()?;
+
+		// Step 2: Fetch all pending transactions
+		let cf_pending = self.db.cf_handle(cf::PENDING_TRANSACTIONS).ok_or_else(|| {
+			DaSequencerError::Generic("Missing column family: pending_transactions".into())
+		})?;
+
+		let mut txs = Vec::new();
+		let iter = self.db.iterator_cf(&cf_pending, rocksdb::IteratorMode::Start);
+		let mut batch = WriteBatch::default();
+
+		for item in iter {
+			let (key, value) = item.map_err(|e| DaSequencerError::Generic(e.to_string()))?;
+
+			let tx: FullNodeTx = bincode::deserialize(&value).map_err(|e| {
+				DaSequencerError::Generic(format!("Tx deserialization failed: {}", e))
+			})?;
+
+			txs.push(tx);
+			batch.delete_cf(&cf_pending, &key); // remove included txs from pending
+		}
+
+		// No transactions to include
+		if txs.is_empty() {
+			return Ok(None);
+		}
+
+		let tx_set: BTreeSet<Transaction> = txs.into_iter().collect();
+
+		// Construct and validate the block
+		// Check that Id:default() generates a unique id every time.
+		let block = Block::new(BlockMetadata::default(), Id::default(), tx_set);
+
+		let sequencer_block = SequencerBlock::try_new(next_height, block)?; // validates size internally
```

Comment:
This will fail if we have more data than the size limit. However, the limit is not enforced when pending transactions are written, is it?

Reply:
Will handle this logic in #1159